lib/opal/nodes/top.rb

```diff
@@ -41,11 +47,74 @@ def compile
           line body_code
         end
 
+        if compiler.es6_modules?
+          import_lines = compiler.requires.map do |module_path|
+            # modules should have the ending .rb for imports, so that the opal-webpack-resolver-plugin
+            # or the opal-webpack-loader don't mix them up with javascript imports
+            # that is just a sensible convention for example, a require 'react', without ending:
+            #   import 'react';  --> resolves to react.js in javascript space, may be resolved by webpack otherwise
+            # vs. with ending:
+            #   import 'react.rb';  --> resolves to react.rb in the opal/ruby space which gets transpiled to js by the loader
+            # if a javascript file gets required as:
+            #   require 'runtime'
+            # it gets imported like so:
+            #   import 'runtime.rb'
+            # the opal-webpack-resolver-plugin will then check for a runtime.rb, but also for a runtime.js if the runtime.rb is not found.
```

Comment:
There are too many comments and LOC that are not related to the `TopNode` *directly*. It definitely should be extracted to a separate utility class, `TopNode` is a utility node responsible for compiling a JS code that wraps a single ruby file.

Your code *mostly* does a different thing: it handles paths, does some recursive walking etc. All these calculations should be extract to some module like `ES6Paths` that must be used by this class.